File: solutions/python/dnd-character/dnd_character.py
import random

def modifier(n):
    return (n - 10) // 2


TRAITS = ('strength', 'dexterity', 'constitution', 'intelligence', 'wisdom', 'charisma')


class Character:

    def __init__(self):
        for trait in TRAITS:
            setattr(self, trait, self.ability())

        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        dice = sorted([random.randint(1, 6) for _ in range(4)], reverse=True)
        return sum(dice[:3])

File: solutions/ruby/dnd-character/dnd_character.rb
class DndCharacter
  attr_accessor :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints

  def self.modifier(attrib)
    (attrib - 10) / 2
  end

  def roll_attrib
    (1..4).to_a.map { Random.rand(1..6) }.sort[1, 3].sum
  end

  def initialize
    self.strength = roll_attrib
    self.dexterity = roll_attrib
    self.constitution = roll_attrib
    self.intelligence = roll_attrib
    self.wisdom = roll_attrib
    self.charisma = roll_attrib
    self.hitpoints = 10 + DndCharacter.modifier(constitution)
  end
end
